Set/Get non-blocking mode via fcntl() on cc3220sf_launchxl

vanti opened the pull request:
This patch set adds support for setting/getting non-blocking mode on cc3220sf_launchxl. In particular, it does the following:

- Revert changes in commit  55b3f059329d87c40e40d82048ecb4232f76021e, given build errors are seen when fcntl.h is included. The latter declares fcntl() as a non-static function. The same function cannot be declared as both static and non-static. @rlubos now I understand why you implemented things the way you did previously.

-   Add additional workarounds to cc_pal.h that are necessary due to the
    fact that the POSIX subsys does not integrate well with other parts of
    Zephyr, as per discussion in #13444. These workarounds are needed when
    fcntl.h is included, to avoid symbols from being redefined in POSIX
    headers. They disable some POSIX headers and define what is necessary
    for the code to compile.

- Implement the fcntl() function in the SimpleLink Wifi driver

- Update the echo_async example to build and run, to validate the change on cc3220sf_launchxl. Because  IPv6 is not currently enabled in the driver, we have to add guards to avoid handling IPv6 in the example.

vanti committed:
drivers: wifi: simplelink: Implement fcntl to set/get non-blocking mode

This commit implements fcntl() in the SimpleLink Wifi driver to set and
get the non-blocking mode on a socket.

Fixes #11891.

Signed-off-by: Vincent Wan <vincent.wan@linaro.org>

vanti committed:
net: sockets: move fcntl back to socket_offload.c

We are reverting the changes in commit
55b3f059329d87c40e40d82048ecb4232f76021e given build errors are seen
when fcntl.h is included, as it declares fcntl() as a non-static
function. The same function cannot be declared as both static and
non-static.

Instead, we avoid redefining fcntl() in lib/os/fdtable.c specifically
for case of the SimpleLink family, til we have support for the new
socket_op_vtable.

Signed-off-by: Vincent Wan <vincent.wan@linaro.org>

vanti committed:
ext: simplelink: add workarounds when using POSIX headers

Adding additional workarounds to cc_pal.h that are necessary due to the
fact that the POSIX subsys does not integrate well with other parts of
Zephyr, as per discussion in #13444. These workarounds are needed when
fcntl.h is included, to avoid symbols from being redefined in POSIX
headers. They disable some POSIX headers and define what is necessary
for the code to compile.

We should aim to remove these when the integration story has improved.

Signed-off-by: Vincent Wan <vincent.wan@linaro.org>

GAnthony commented on drivers/wifi/simplelink/simplelink_sockets.c:
```diff
@@ -976,14 +977,36 @@ static void simplelink_freeaddrinfo(struct addrinfo *res)
+
+	switch (cmd) {
+	case F_GETFL:
+		sl_GetSockOpt(sd, SL_SOL_SOCKET, SL_SO_NONBLOCKING,
```
Perhaps, here and below, check the return value of sl_[Get/Set]SockOpt(),  then pass retval via
`slcb_SetErrno(retval);`
as in the other cases.

vanti replied:
Thanks. I will make the change.

GAnthony commented on drivers/wifi/simplelink/simplelink_sockets.c:
```diff
@@ -976,14 +977,36 @@ static void simplelink_freeaddrinfo(struct addrinfo *res)
 
-	errno = ENOTSUP;
-	return -1;
+	return setErrno(retval);
```
`slcb_SetErrno(retval);`  ?

vanti replied:
setErrno() calls slcb_SetErrno() after converting sl_ error codes into standard error codes.

GAnthony reviewed (commented):
Looks good.
A couple minor comments

jukkar commented on lib/os/fdtable.c:
```diff
@@ -219,6 +219,12 @@ int ioctl(int fd, unsigned long request, ...)
+ * feature. The socket offload subsys has already defined fcntl, hence we
+ * avoid redefining fcntl here.
+ */
+#ifndef CONFIG_NET_SOCKETS_OFFLOAD
```
The fcntl() is generic function that should be available even if we have defined CONFIG_NET_SOCKETS_OFFLOAD. Now you are preventing the use of it for those file descriptors that are not offloaded.

vanti replied:
Changed to making it SimpleLink-specific.

jukkar commented on samples/net/sockets/echo_async/src/socket_echo.c:
```diff
@@ -102,19 +102,23 @@ int main(void)
 {
```
You need to split the "samples: sockets: echo_async: Add support for cc3220sf_launchxl " patch. The board specific changes should go to a patch of its own.

vanti replied:
Thanks, I made the split.

jukkar commented on samples/net/sockets/echo_async/src/socket_echo.c:
```diff
@@ -102,19 +102,23 @@ int main(void)
 			.s_addr = htonl(INADDR_ANY),
 		},
 	};
+#ifdef CONFIG_NET_IPV6
```
The sample can be used both in Zephyr and Linux, but Linux does not know anything about CONFIG_NET_IPV6. So this needs to be changed here.

pfalcon replied:
Yeah, in a sense that adding this #ifdef would blindly disable it for Linux (or another POSIX OS). I'd suggest to use something like `#ifdef USE_IPV6`, and set USE_IPV6 in the header of this file, where Zephyr-vs-POSIX ifdef already is.

vanti replied:
Thanks, done.

pfalcon commented on samples/net/sockets/echo_async/src/socket_echo.c:
```diff
@@ -143,14 +149,18 @@ int main(void)
-
 	pollfds_add(serv4);
+
+#ifdef CONFIG_NET_IPV6
```
Would be nice to minimize number of #ifdef's, and arrange complete setup of ipv4 socket, then complete setup of ipv6 socket (wrapped in #ifdef).

vanti replied:
Done

vanti committed:
samples: sockets: echo_async: Make IPV6 optional

Changing the example to not use IPv6 when it is not configured,
given IPv6 is not enabled on some platforms such as cc3220sf_launchxl.

Signed-off-by: Vincent Wan <vincent.wan@linaro.org>

vanti committed:
samples: sockets: echo_async: Add support for cc3220sf_launchxl

Adding support for cc3220sf_launchxl by introducing a board-specific
Kconfig file.

Signed-off-by: Vincent Wan <vincent.wan@linaro.org>

vanti commented:
Updated based on comments from @pfalcon and @jukkar, and also modified error handling in simplelink_fcntl() to match the new style from #14285 which was recently merged.

pfalcon commented on lib/os/fdtable.c:
```diff
@@ -227,6 +227,12 @@ int ioctl(int fd, unsigned long request, ...)
+ * feature. The socket offload subsys has already defined fcntl, hence we
+ * avoid redefining fcntl here.
+ */
+#ifndef CONFIG_SOC_FAMILY_TISIMPLELINK
```
This is misplaced here. This file is not SoC-specific, so should not contain stuff like that. I'm going to approve this PR, but please work towards cc32xx compatibility with Zephyr POSIX subsys (this #ifdef effectively disables it), and implementation of fcntl via fdtable->ioctl().

jukkar replied:
I very much agree with @pfalcon here. This is just in wrong place.
Aren't you installing `simplelink_fcntl` function in simplelink_sockets.c so what is the purpose of this stuff here?

vanti replied:
@pfalcon Yes I agree that ultimately for cc32xx to conform with the flow of using fdtable is probably the right thing to do in the long run. This is only done as a short term solution for fcntl support in 1.14.

@jukkar I had to resort to taking this code out for SimpleLink because there needs to be another definition of fcntl() in subsys/net/lib/sockets/socket_offload.c which generates the energy to call simplelink_fcntl(). Originally I wanted to do this for all cases where NET_SOCKET_OFFLOAD is set, but to your earlier point that'd prevent others who have enabled fdtable support in offloaded sockets to benefit from unified file descriptor handling. So given where we are with SimpleLink today (ie. no fdtable support) and what was agreed upon in #11891 , this made the most sense to me in order to prevent fcntl() from being defined twice, and to be able to continue to follow a separate path for all offloaded functions.

jukkar replied:
@vanti Ok, I suppose we can leave it like this until a proper support is added.

pfalcon reviewed (approved):
Thanks for the sample changes.

GAnthony commented:
@galak, ready to merge.
